### ISSUE TITLE
Tackling Fairness Issue

### ISSUE THREAD

**Issue by tbssajal** [autonity/autonity#860]
This issue describes how the fairness issue #819 can be triggered without 100% slashing and discusses steps taken to solve the issue.
**_NOTE: In all calculations the unit of stake is wei unless mentioned otherwise_**

In simple terms, fairness issue can be triggered if `delegatedStake` or `unbondingStake` become 0 from positive due to slashing. I will be descibing how the current mechanism can trigger fairness issue without slashing 100% when both `delegatedStake > 0 and unbondingStake > 0`.

# Triggering The Issue
Let's consider we have `selfBondingStake = 0` and `selfUnbondingStake = 0` for simplicity. It doesn't have to be the case to trigger the issue but it is easy to discuss. So we have `totalStake = delegatedStake + unbondingStake`. The current slashing mechanism divides the slashing amount proportionally to the stake amount. If we slash a total `slashingAmount` from `totalStake`, then we slash the following from `delegatedStake`:
`delegatedSlash = (slashingAmount * delegatedStake) / totalStake`   -----> eqn1[^eqn1]
Similarly we slash the following amount from `unbondingStake`:
`unbondingSlash = (slashingAmount * unbondingStake) / totalStake` ------> eqn2[^eqn2]
After calculating the slashing amount from above equations, if we have `delegatedSlash == delegatedStake and delegatedStake > 0` we have triggered fairness issue. Same goes for `unbondingSlash`. We can trigger it by slashing 100% as we already know, but we can also trigger it accidentally without 100% slashing due to integer division. In current implementation, we always take floor for `unbondingSlash` and ceil for `delegatedSlash`. So only `delegatedSlash` can be equal to `delegatedStake` without 100% slashing.

## Conditions to Trigger
We can do some math to find the condition to trigger the issue without 100% slashing.

<details>
<summary>
Click to see details.
</summary>

Let's denote,
$delegatedStake = d,$
$slashingAmount = s,$
$totalStake = t,$
$delegatedSlash = d^{\prime}$

We have the following:
$d^{\prime} = ceil\left(sd/ t\right)$
$or, d^{\prime} = floor\left((sd + t - 1) / t\right)$

Using inequality and putting $d^{\prime} = delegatedSlash = delegatedStake = d$ because we want to trigger fairness issue,
$d \leq (sd + t - 1) / t$
$or, dt \leq sd + t - 1$
$or, d \leq (t - 1) / (t - s)$ -----> eqn3[^eqn3]

So if we have $d \leq (t - 1) / (t - s)$ then we are in trouble and the upper limit of the trouble is $d_u = floor((t - 1) / (t - s))$. We will be in trouble if $delegatedStake \leq d_u$ and therefore we want to minimize $d_u$.

Let's replace $slashingAmount = s = rt$ where $r = slashing \\ rate < 1$. Now we have
$d \leq (t - 1) / (t * (1 - r))$
$or, d \leq (1 - 1/t) / (1 - r)$

For simplicity let's denote a factor $x$ as:
$x = 1/(1-r)$ -------> eqn4[^eqn4]

Now we can write the upper limit $d_u$ as
$d_u = floor((1 - 1/t) * x) = floor(x - x/t)$ ------> eqn5[^eqn5]

Note that $x \leq t$ because $x = 1/(1-r)$ and $r = slashingAmount / totalStake$, even if we use very high precision, $x \leq t$ will be true.
</details>

Note that we can actually write:

$remainingDelegatedStake = floor(delegatedStake / x)$ [Because we are taking ceil while calculating $delegatedSlash$]
and,
$remainingUnbondingStake = ceil(unbondingStake / x)$ [Because we are taking floor while calculating $unbondingSlash$]
where
$x = 1/(1-r)$ where $r = slashingRate < 1$. And we get the upper limit of $delegatedStake$ to trigger the fairness issue, $d_u = floor(x - x/t)$

## Example
Let me put an example just for the perspective, if we have 1 million NTN stakes, i.e. $t = 10^{24} \\ wei$, consisting only of delegated stake and unbonding stake and if we support the best precision, then it is possible to get $x = 10^{20}$ even if it is highly unlikely. Then we get $d_u = 10^{20} - 1$ which is less than 100 NTN. So we have the following:

$remainingDelegatedStake = floor(delegatedStake / x)$ where $delegatedStake < 100 \\ NTN \leq 10^{20} - 1$
so,
$remainingDelegatedStake = floor((10^{20} - 1) / x) = 0$

That means even with delegated stake less than 100 NTN, there is a possibility to trigger fairness issue and if triggered we will slash all 100 NTN of delegated stakes but will have a lot of unbonding stakes remaining.

$remainingUnbondingStake = ceil(unbondingStake / x)$ where $unbondingStake > 10^6 - 100 \\ NTN$
so,
$remainingUnbondingStake = 10000 \\ NTN$

# Solution
Discussing three solutions. The first two solutions are depending on the fact that we can control slasing rate $r$ and some observations and assumptions. The third one is more elegant in my opinion but introduces additional logic.

## First Solution
If we control $slashingRate, \\ r \leq 0.9$ then we have $x \leq 10$. That means the fairness issue can only be triggered with $d_u < 10 \\ wei$. Even though we can still have delegated stake < 10 **wei** and unbonding stake in million **NTN** (notice wei and NTN), we can assume it is highly unlikely for user to request unbonding of millions of NTN and keeping less than 10 wei bonded. And if we assume delegated stake was very high at some point and got low to 10 wei by slashing only, then also it would have to be slashed very repeatedly and very roughly. So this is highly unlikely to happen but doing a lot of assumptions here. Also we cannot do 100% slashing here as we are controlling $slashingRate, \\ r \leq 0.9$. But I am not sure how we decide to `jailbound`[^kill] the validator in this case.

## Second Solution
We control the precision of slashing rate, i.e. how many digits we use. For example right now we have $slashingRatePrecision = 10000$, so we are using 4 digits and therefore $slashingRate, \\ r \leq 0.9999$ and we have $x \leq 10000$. So we will always have $x \leq slashingRatePrecision$. So we have the same conclusion as above because $x$ is very low. We can assume that nobody will keep only 10000 wei bonded and unbond all the rest. However we can do 100% slashing in this case unlike the first solution. But again I am not sure how to decide to `jailbound`[^kill] the validator in this case as well.

## Third Solution
As done in PR #857 we modify the logic a little bit. Instead of taking the ceil for $delegatedSlash$ and floor for $unbondingSlash$ we take floor for the smaller one and ceil for the larger one. So that means we will take ceil for $delegatedSlash$, only `if (delegatedStake >= unbondingStake)` and in that case we will have $delegatedStake \geq totalStake / 2$.

If we do the math again.

<details>
<summary>
Click to see details.
</summary>

We can put $d \geq t/2$ in eqn3[^eqn3], so we have
$t/2 \leq d \leq (t-1)/(t-s)$
$or, s \geq t-2+2/t$
$or, s \geq t-1$ [as all of them are integers]

</details>

We have $slashingAmount \geq totalStake - 1$. That means we can say that to trigger fairness issue, we will have to slash at least $totalStake - 1$ or the $totalStake$. Although we are introducing the new logic and it may be a little complex on the code, it ensures that unless all the stake is reduced to 1 or 0 wei we will not have fairness issue.

## Notes on The Solutions
For the first and second solutions, as we are relying on $slashingRate$ or $slashingRatePrecision$, I think it could be troublesome as it depends on our understanding how to change them. For example if someone changes the parameters $slashingRate$ or $slashingRatePrecision$ by mistake in future, the first two solutions will not work. Also it takes away the freedom to tune the parameters as the solutions will put some constraints on them. So I prefer the third solution. Also for the third one it is easy to determine when to `jailbound`, we can `jailbound` if $slashingAmount \geq totalStake - 1$. Please comment if anyone has any alternative or anything that I missed in above solutions.

# Summary
So basically it is possible to trigger fairness issue without 100% slashing following the eqn3[^eqn3]. As we must `jailbound` validator on triggering fairness issue, this can happen even if there are a lot of stakes remaining as `unbondingStake` at least mathematically. Furthermore, it is difficult to decide on the rules when to `jailbound`. I suggest we adopt a little change in the logic to ensure that no stakes are remaining when fairness issue is triggered. So we don't have to `jailbound` a validator when he has a lot of stakes.

[^eqn1]: from [codebase](https://github.com/autonity/autonity/blob/develop/autonity/solidity/contracts/Accountability.sol#L381)
[^eqn2]: from [codebase](https://github.com/autonity/autonity/blob/develop/autonity/solidity/contracts/Accountability.sol#L378)
[^eqn3]: limit on delegated stake
[^eqn4]: factor related to slashing rate
[^eqn5]: delegated stake upper limit
[^kill]: ban permanently

**_Edit1: On the example_**
**_Edit2: Hiding the long math and adding the conclusion to skip the math_**

**Comment by lorenzo-dev1** [comment]
Awesome work.

Couple comments, probably due to myself not understanding some math passages:
- might be worth adding a reference to explain the conversion from ceil to floor (e.g. https://math.stackexchange.com/questions/448300/convert-ceil-to-floor)
- A few clarifications regarding eqn3. If we assume that $/$ does integer division with floor rounding like in solidity, correct me if I am wrong:
  -  $d \lt (t - 1) / (t - s)$ --> cannot happen because it would mean that we are slashing > 100% (?)
  - $d = (t - 1) / (t - s)$ --> fairness issue is triggered (we slashed all the delegated stake)
  - $d \gt (t - 1) / (t - s)$ --> no fairness issue, we slashed only a part of the delegated stake.
- for eqn.5, I think that $x \leq t$ is imprecise, it should be $x \lt t$. The case $x = t$ is possible only if we are slashing 100% (which we are excluding for this computation).
-  still for eqn.5, why $d_u = x-1$ ? If we set $x \rightarrow t$ in $d_u = floor(x - x/t)$ we obtain: $d_u \rightarrow floor(t - 1)$. Not sure how the floor plays into this equation though...
- How did you derive the formulas for $remainingDelegatedStake$ and $remainingUnbondingStake$

I will comment more on the solutions once I fully understood this math passages, however option 3 seems the most complete to me as it solves the problem without any assumption on the stake distribution.

On a side note:
- What happens if we slash a node with total stake $t = 0$ ? Maybe we already taken care of this in the code, but worth double checking.

**Comment by tbssajal** [comment]
> might be worth adding a reference to explain the conversion from ceil to floor

Thanks for the reference Lorenzo! Actually I did not know that the reference existed or easily accessible, but I also did not put any effort to find it.

> If we assume that / does integer division with floor rounding like in solidity

When doing math, I am assuming '/' does normal division, not integer division. Because I don't know how to apply mathematical operations assuming integer division. I am writing "floor" when I mean integer division.

> $d < (t-1)/(t-s)$ cannot happen because it would mean that we are slashing > 100%

I am not sure what you mean. Pick any amount of $totalStake, \\ t$ and $slashingAmout, \\ s < t$. Then if we pick any amount of $delegatedStake, \\ d < (t-1)/(t-s)$ and $unbondingStake, \\ u = t - d$ and then apply the slashing, $s$ on them accordingly we can trigger the fairness issue. The other two points are correct.

> for eqn.5, I think that $x \leq t$ is imprecise

I didn't think about it. I could be true. Either way, I think it doesn't matter much. But thanks for pointing out.

> still for eqn.5, why $d_u = x-1$ ? If we set $x \rightarrow t$ in $d_u = floor(x - x/t)$ we obtain: $d_u = floor(t-1)$. Not sure how the floor plays

Actually I should have wrote $d_u = floor(x-1)$. Anyway consider $d_u$ just an upper limit here. $d_u$ could be fraction, it just means if $d \leq d_u$ then we have fairness issue. So yeah if $x \rightarrow t$ then we can trigger fairness issue with $d_u = floor(t-1)$. So with $delegatedStake, \\ d \leq t-1$ fairness issue can be triggered. It is because $x \rightarrow t$ means a very high precision and the highest amount of slashing possible which is less than 100%.

> How did you derive the formulas for $remaningDelegatedStake$

Sorry I edited it later so I rushed a little. So we can write mathematically
$remaningDelegatedStake = delegatedStake - delegatedSlash$
$or, \ remaningDelegatedStake = delegatedStake - delegatedStake * slashingAmount / totalStake$
$or, \ remaningDelegatedStake = delegatedStake * (1-slashingRate)$
$or, \ remaningDelegatedStake = delegatedStake * (1-r)$ [ putting $r = slashingRate$ ]

and we have $x = 1/(1-r)$ so we can write $remaningDelegatedStake = delegatedStake / x$. But in solidity we take ceil for $delegatedSlash$. As it is subtracted from $delegatedStake$ to get $remaningDelegatedStake$, we can write $remaningDelegatedStake = floor(delegatedStake / x)$.

Same goes for $remaningUnbondingStake$.

> What happens if we slash a node with total stake $t = 0$ ?

We already handle it because we calculate $slashingAmount, \\ s$ first. If we get $t = 0$, after slashing $selfbondedStake$ and $selfUnbondingStake$, we will get remaining $slashingAmount, \\ s = 0$.

**Comment by lorenzo-dev1** [comment]
> When doing math, I am assuming '/' does normal division, not integer division. Because I don't know how to apply mathematical operations assuming integer division. I am writing "floor" when I mean integer division.

Ok that's actually the correct idea

> I am not sure what you mean. Pick any amount of ...

So it is possible that $delegatedSlash > delegatedStake$ ? That's what that formula implies if I understood correctly.

> Actually I should have written ...

My point is: if you substitute $x/t$ with $1$, then you also have to substitute $x$ with $t$. If you do not assume $x \rightarrow t$, how do you substitute $x/t$ with $t$?

> Sorry I edited it later so I rushed a little. So we can write mathematically

Got it. So $r$ is the cumulative slashing rate (already taking into account history and collusion), right?

**Comment by lorenzo-dev1** [comment]
@tbssajal I checked the math in the solutions and it seems solid to me. I confirm that I believe that the Solution 3 is the most comprehensive and efficient solution.

Just for verification I tried to derive the conditions for the triggering of the fairness issue on the $unbondingStake$ (remaining in our example where we take ceil of $delegatedStake$ and floor of $unbondingStake$). We should have:

$$ u' = floor(\frac{su}{t})$$

$$ u \leq \frac{su}{t}$$

$$ t \leq s$$

Which since we cannot slash >100%, becomes:

$$ t = s$$

Which is the only condition (100% slash) where we can have the fairness issue in the "flooring" side.

@tbssajal Please let me know if this passages make sense to you.

**Comment by tbssajal** [comment]
> Got it. So $r$ is the cumulative slashing rate (already taking into account history and collusion), right?

Yes.

> Which is the only condition (100% slash) where we can have the fairness issue in the "flooring" side.

Correct. The flooring side will never reach zero unless slashed 100%.

**Comment by yazzaoui** [comment]
Thank you @tbssajal for this detailed write-up. Following our call my first suggestion was to change the `delegatedSlash` formula to use a flooring method, e.g.:
```solidity
uint256 _delegatedSlash = (_remaining * (_val.bondedStake - _val.selfBondedStake)) /
                                        (_val.unbondingStake + (_val.bondedStake - _val.selfBondedStake));
```

But after reflection, I think that would still cause the issue in case of a 100% slash. 

IMO the easiest and the most sound solution would be something close to option 1. If the slashing rate is above a protocol parameter R, which must be below 99%, then we `jailbound` the validator with a 100% slash. Not only does this fix this issue but from a security standpoint it makes also complete sense.

There is another issue that is not covered here, it's about slashing multiple times at a rate close to 1, like 0.999999999, which may cause an integer overflow issue in the stake supply.

**Comment by tbssajal** [comment]
> Following our call my first suggestion was to change the delegatedSlash formula to use a flooring method, e.g.:
> ```
> uint256 _delegatedSlash = (_remaining * (_val.bondedStake - _val.selfBondedStake)) /
>                                        (_val.unbondingStake + (_val.bondedStake - _val.selfBondedStake));
> ```
> But after reflection, I think that would still cause the issue in case of a 100% slash.

The issue will always be triggered with 100% slash, no matter whatever approach is taken. So in 100% slash we can jailbound the validator. If we change `_delegatedSlash` formula to use floor, the issue can not be triggered unless slashed 100%. So I vote for this approach. It does not make the calculation too complex.

> IMO the easiest and the most sound solution would be something close to option 1. If the slashing rate is above a protocol parameter R, which must be below 99%, then we `jailbound` the validator with a 100% slash

99% or 90%? Just to be clear, even in this case there is a chance to trigger the issue if someone has delegated stake less than 10 wei, which is very unlikely and we don't care about it.

> There is another issue that is not covered here, it's about slashing multiple times at a rate close to 1, like 0.999999999, which may cause an integer overflow issue in the stake supply.

Integer overflow due to ceil or floor in `_delegatedSlash` or `_unbondingSlash` calculation? I don't think it will happen because we always have `_remaining <= totalStake` in the formula. So we will always have `_delegatedSlash <= delegatedStake` and `_unbondingSlash <= unbondingStake` true, no matter which one is ceil or which one is floor or even both are floor or ceil at the same time. But maybe I am missing your point?

**Comment by lorenzo-dev1** [comment]
So is the opinion to not have the `if` that checks which of the two stake parts is lower, and therefore which one we need to floor?

**Comment by tbssajal** [comment]
> @yazzaoui @tbssajal is the opinion to remove the `if` that checks which of the two stake parts is lower, and therefore which one we need to floor? And always floor delegatedStake?
> 
> If so, why? I understand that it removes one if (lower code complexity), but that is at the cost of having an imperfect solution rather than perfect one.

Yeah always floor both `_delegatedSlash` and `_unbondingSlash`. This approach reduces any additional complexity. In some case we may slash 1 wei less than the amount supposed to be slashed, which is not a big deal I guess. Actually I am fine with this one or the third solution that I suggested in the original post, i.e. taking floor for the one which is smaller and ceil for the other one.

**Comment by yazzaoui** [comment]
> The issue will always be triggered with 100% slash, no matter whatever approach is taken. So in 100% slash we can jailbound the validator. If we change _delegatedSlash formula to use floor, the issue can not be triggered unless slashed 100%. So I vote for this approach. It does not make the calculation too complex.

Alright, then if everyone agree let's move on with this solution.

> 99% or 90%? Just to be clear, even in this case there is a chance to trigger the issue if someone has delegated stake less than 10 wei, which is very unlikely and we don't care about it.

Well, a malicious actor can then trigger it by bonding a precise amount and get slashed accordingly. So I guess in this scenario we need the above agreed solution. But I do still think that we need this threshold mechanism set in place. 

 > But maybe I am missing your point?
 
 The idea is that by slashing at 99.999999% you create a disproportionate Share/Token ratio (or LNTN/NTN conversion rate) so we end up with situations where 10000....0000 LNTN is equal to 1 NTN

**Comment by tbssajal** [comment]
> The idea is that by slashing at 99.999999% you create a disproportionate Share/Token ratio (or LNTN/NTN conversion rate) so we end up with situations where 10000....0000 LNTN is equal to 1 NTN

I see. I think it could be a separate issue.

**Comment by tbssajal** [comment]
Resolved by #857
